File: solutions/news_distribution.c
/**
 * tags: classic, dsu, dfs, graphs
 * statement: https://codeforces.com/problemset/problem/1167/C
*/


#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>


int n, m, k, u, v;

int arr[500005];

int find(int i) {
    if (arr[i] < 0) return i;
    arr[i] = find(arr[i]);
    return arr[i];
}

void union_sets(int i, int j) {
    int a = find(i);
    int b = find(j);
    if (a == b)
        return;
    if (arr[a] < arr[b]) {
        arr[a] += arr[b];
        arr[b] = a;
    }
    else {
        arr[b] += arr[a];
        arr[a] = b;
   }
}



int main(int argc, char* argv[]) {
    scanf("%d%d", &n, &m);
    for (int i=0; i<n; i++) {
        arr[i] = -1;
    }
    for (int i=0; i<m; i++) {
        scanf("%d", &k);
        if (k > 0) {
            scanf("%d", &u);
            while (--k) {
                scanf("%d", &v);
                union_sets(u-1, v-1);
            }
        }
    }
    for (int i=0; i<n; i++) {
        printf("%d ", -arr[find(i)]);
    }
    return 0;
}

File: solutions/NewsDistribution.kt
/**
 * tags: graphs, dsu, dfs, classic
 * statement: https://codeforces.com/problemset/problem/1167/C 
*/

import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.util.StringTokenizer



fun main() {
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    val m = sc.nextInt()
    val dsu = Array(n) {-1}
    for (i in 0..<m) {
        val k = sc.nextInt()
        if (k > 0) {
            val a = sc.nextInt() - 1
            for (j in 0..<k - 1) {
                union(dsu, a, sc.nextInt() - 1)
            }
        }
    }
    for (i in 0..<n) {
        print("${-dsu[find(dsu, i)]} ")
    }
}

fun find(arr: Array<Int>, i: Int): Int {
    if (arr[i] < 0) return i
    arr[i] = find(arr, arr[i])
    return arr[i]
}

fun union(arr: Array<Int>, i: Int, j: Int) {
    val a = find(arr, i)
    val b = find(arr, j)
    if (a != b) {
        if (arr[a] < arr[b]) {
            arr[a] = arr[a] + arr[b]
            arr[b] = a
        }
        else {
            arr[b] = arr[b] + arr[a]
            arr[a] = b
        }
    }
}

class Scanner(private val inputStream: InputStream) {
    private var st: StringTokenizer? = null
    private val br: BufferedReader = BufferedReader(InputStreamReader(inputStream))

    fun next(): String {
        while (st == null || !st!!.hasMoreTokens()) {
            st = StringTokenizer(br.readLine())
        }
        return st!!.nextToken()
    }
    fun nextInt() = next().toInt()
}
